lib/paperclip/color_extractor.rb

```diff
@@ -0,0 +1,174 @@
+# frozen_string_literal: true
+
+require 'mime/types/columnar'
+
+module Paperclip
+  class ColorExtractor < Paperclip::Processor
+    MIN_CONTRAST = 3.0
+
+    def make
+      depth = 8
+
+      # Determine background palette by getting colors close to the image's edge only
+      background_palette = palette_from_histogram(convert(':source -crop :crop -format %c -colors :quantity -depth :depth histogram:info:', source: File.expand_path(@file.path), quantity: 4, depth: depth, crop: '50%x100%'), 4)
```

Comment:
I'm not sure how this achieves “getting colors close to the image's edge only? From my understanding, it builds two histograms, of the two halves of the image…

Reply:
One is of the left half, one is of the full image. Couldn't do anything better.

Reply:
No! `-crop 50%x100%` actually processes two images, the two halves! You want `-crop 50%x100%+0+0` for only the left half.

Reply:
Okay, good catch -- this led me down a rabbit hole that uncovered quite a few mistakes in the code.